core/store/src/trie/state_parts.rs

```diff
@@ -73,20 +78,158 @@ impl Trie {
+        let recorded_trie = recording_trie.recorded_storage().unwrap();
+        let PartialState::TrieValues(path_boundary_nodes) = recorded_trie.nodes;
+
+        // 2. Extract all key-value pairs in state part from flat storage.
+        let values_read_start = Instant::now();
+        let flat_state_iter = self.iter_flat_state_entries(nibbles_begin, nibbles_end)?;
+        let mut values_ref = 0;
+        let mut values_inlined = 0;
+        let all_state_part_items: Vec<_> = flat_state_iter
+            .map(|(k, v)| {
+                let value = match FlatStateValue::try_from_slice(&v).expect("Borsh cannot fail") {
+                    FlatStateValue::Ref(value_ref) => {
+                        values_ref += 1;
+                        self.storage.retrieve_raw_bytes(&value_ref.hash).map(|bytes| bytes.to_vec())
+                    }
+                    FlatStateValue::Inlined(value) => {
+                        values_inlined += 1;
+                        Ok(value)
+                    }
+                };
+                value.map(|v| (k, Some(v)))
+            })
+            .collect::<Result<_, _>>()?;
+        let values_read_duration = values_read_start.elapsed();
+
+        // 3. Create trie out of all key-value pairs.
+        let trie_creation_start = Instant::now();
+        let local_state_part_trie =
+            Trie::new(Rc::new(TrieMemoryPartialStorage::default()), StateRoot::new(), None);
+        let local_state_part_nodes =
+            local_state_part_trie.update(all_state_part_items.into_iter())?.insertions;
+
+        // 4. Unite all nodes in memory, traverse trie based on them, return set of visited nodes.
+        let boundary_nodes_storage: HashMap<_, _> =
+            path_boundary_nodes.iter().map(|entry| (hash(entry), entry.clone())).collect();
+        let disk_read_hashes: HashSet<_> = boundary_nodes_storage.keys().cloned().collect();
+        let mut all_nodes: HashMap<CryptoHash, Arc<[u8]>> = HashMap::new();
+        all_nodes.extend(boundary_nodes_storage);
+        all_nodes.extend(
+            local_state_part_nodes
+                .iter()
+                .map(|entry| (*entry.hash(), entry.payload().to_vec().into())),
+        );
+        let final_trie =
+            Trie::new(Rc::new(TrieMemoryPartialStorage::new(all_nodes)), self.root, None);
+
+        final_trie.visit_nodes_for_state_part(part_id)?;
+        let final_trie_storage = final_trie.storage.as_partial_storage().unwrap();
+        let final_state_part_nodes = final_trie_storage.partial_state();
+        let PartialState::TrieValues(trie_values) = &final_state_part_nodes;
+
+        // Compute how many nodes were recreated from memory.
+        let state_part_num_nodes = trie_values.len();
+        let in_memory_created_nodes =
+            trie_values.iter().filter(|entry| !disk_read_hashes.contains(&hash(*entry))).count();
+        let trie_creation_duration = trie_creation_start.elapsed();
+        tracing::info!(
```

Comment:
I suggest to consider avoiding manually formatting string for logging in general and use structured logging instead: `%values_ref, ..., ?trie_creation_duration, ...`. This s a good practice because it still results in very readable logs, but avoids unnecessary string formatting overhead when log level is not enabled. Also I find that to be a bit easier to `grep` in the logs. Not a strong opinion here, just something to consider.